diff --git a/osfmk/arm64/machine_routines.c b/osfmk/arm64/machine_routines.c
index 3ed62bd6..f1d0d190 100644
--- a/osfmk/arm64/machine_routines.c
+++ b/osfmk/arm64/machine_routines.c
@@ -2097,7 +2097,7 @@ static inline uint64_t
 nonspeculative_timebase(void)
 {
 #if defined(HAS_ACNTVCT)
-	return __builtin_arm_rsr64("ACNTVCT_EL0");
+	return __builtin_arm_rsr64("S3_4_c15_c10_6");
 #elif __ARM_ARCH_8_6__
 	return __builtin_arm_rsr64("CNTVCTSS_EL0");
 #else
