diff --git a/osfmk/vm/vm_map.c b/osfmk/vm/vm_map.c
index 5e2209436..826c86842 100644
--- a/osfmk/vm/vm_map.c
+++ b/osfmk/vm/vm_map.c
@@ -14294,15 +14294,15 @@ protection_failure:
 			fault_info->cs_bypass = FALSE;
 		}
 		fault_info->pmap_cs_associated = FALSE;
-#if CONFIG_PMAP_CS
-		if (entry->pmap_cs_associated) {
-			/*
-			 * The pmap layer will validate this page
-			 * before allowing it to be executed from.
-			 */
-			fault_info->pmap_cs_associated = TRUE;
-		}
-#endif /* CONFIG_PMAP_CS */
+// #if CONFIG_PMAP_CS
+// 		if (entry->pmap_cs_associated) {
+// 			/*
+// 			 * The pmap layer will validate this page
+// 			 * before allowing it to be executed from.
+// 			 */
+// 			fault_info->pmap_cs_associated = TRUE;
+// 		}
+// #endif /* CONFIG_PMAP_CS */
 		fault_info->mark_zf_absent = FALSE;
 		fault_info->batch_pmap_op = FALSE;
 		fault_info->resilient_media = entry->vme_resilient_media;
