diff --git a/EXTERNAL_HEADERS/CodeSignature/Entitlement.h b/EXTERNAL_HEADERS/CodeSignature/Entitlement.h
new file mode 100644
index 00000000..c91fb061
--- /dev/null
+++ b/EXTERNAL_HEADERS/CodeSignature/Entitlement.h
@@ -0,0 +1,13 @@
+#ifndef CodeSignature_Entitlements_h
+#define CodeSignature_Entitlements_h
+
+__BEGIN_DECLS
+/* ref: xnu/tests/entitlements/hardened-binary-entitlements-2.plist */
+
+#define kCSWebBrowserHostEntitlement "com.apple.private.web-browser-engine.host"
+#define kCSWebBrowserGPUEntitlement "com.apple.private.web-browser-engine.gpu"
+#define kCSWebBrowserNetworkEntitlement "com.apple.private.web-browser-engine.network"
+#define kCSWebBrowserWebContentEntitlement "com.apple.private.web-browser-engine.webcontent"
+
+__END_DECLS
+#endif /* CodeSignature_Entitlements_h */
