diff --git a/pexpert/pexpert/arm64/apple_arm64_common.h b/pexpert/pexpert/arm64/apple_arm64_common.h
index 56caa9556..460c4c4cc 100644
--- a/pexpert/pexpert/arm64/apple_arm64_common.h
+++ b/pexpert/pexpert/arm64/apple_arm64_common.h
@@ -89,11 +89,4 @@
 #define CPU_VERSION_C0                       0x20
 #define CPU_VERSION_UNKNOWN                  0xff
 
-
-/*
- * Conservatively assume that BTI will be enforced.
- * Individual SoCs and kernel configurations may have different behavior.
- */
-#define BTI_ENFORCED 1
-
 #endif /* !_PEXPERT_ARM64_APPLE_ARM64_COMMON_H */
